### PR TITLE
download media

### PR DESCRIPTION
**Pull request by Rits1272**
## Issue that this pull request solves

Closes: #84 

## Proposed changes
This PR will let users download media from one contact

## Types of changes

_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (Documentation content changed)
- [ ] Other (please describe): 

## Checklist

_Put an `x` in the boxes that apply_

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

## Screenshots

Please attach the screenshots of the changes made in case of change in user interface
Screenshot (75)

## Other
How to use: `python -m wplay -wd AnyName` or `python -m wplay --download_media AnyName`
Input number of media to be downloaded.

### CODE REVIEW

**Comment by Rits1272** [comment]
Please review @priyansh19 @xandao6 @rpotter12

File: wplay/download_media.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    # Click on the `Media, Link and Docs` text
+    await page.waitForSelector(media_text)
+    await page.click(media_text)
+
+    # Click on the most recent media element
+    while True:
+        try:
+            await page.evaluate(f'''document.querySelector('{media_images}').click()''')
+            break
+        except:
```
Comment (rpotter12) [inline wplay/download_media.py:59]: Codacy Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- wplay/download_media.py  19
- wplay/__main__.py  1
         

Clones added
============
- wplay/download_media.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163&bid=17848965)

File: wplay/download_media.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                    if vid:
+                        vid = vid[-1]
+                        media_arr['vid'].append(vid)
+                        break
+            
+            # Go to next media element
+            time.sleep(0.5)
+            await page.waitForSelector(left_arrow_button)
+            await page.evaluate(f'''document.querySelector('{left_arrow_button}').click()''')
+
```
Comment (rpotter12) [inline wplay/download_media.py:98]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163)

File: wplay/download_media.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            print("Error saving image")
+
+    for video in media_arr['vid']:
+        try:
+            count += 1
+            viewSource = await newPage.goto(video)
+            f = open(media_path / f'{count}.mp4', 'wb')
+            f.write(await viewSource.buffer())
+            f.close()
+        except:
```
Comment (rpotter12) [inline wplay/download_media.py:122]: Codacy Issue found: [No exception type(s) specified](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163)

Comment (rpotter12) [reply wplay/download_media.py:122]: @Rits1272 add the exception type here

File: wplay/download_media.py (hunk 2 above)
Comment (rpotter12) [reply wplay/download_media.py:98]: @Rits1272 see the issue

File: wplay/download_media.py (hunk 1 above)
Comment (rpotter12) [reply wplay/download_media.py:59]: specify exception type here

File: wplay/download_media.py (hunk 3 above)
Comment (Rits1272) [reply wplay/download_media.py]: Done @rpotter12

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- wplay/download_media.py  19
- wplay/__main__.py  1
         

Clones added
============
- wplay/download_media.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163&bid=17848965)

File: wplay/download_media.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                    img = await page.evaluate(f'''() => [...document.querySelectorAll('{media_url_img}')]
+                                                        .map(element => element.src)''')
+                    if img and len(img) == 2:
+                        img = img[-1]
+                        if img not in media_arr:
+                            media_arr['img'].append(img)
+                        break
+                    countTry += 1
+                    time.sleep(0.3)
+            except:
```
Comment (rpotter12) [inline wplay/download_media.py:83]: Codacy Issue found: [No exception type(s) specified](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163)

File: wplay/download_media.py (hunk 2 above)
Comment (rpotter12) [inline wplay/download_media.py:98]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163)

File: wplay/download_media.py (hunk 1 above)
Comment (rpotter12) [inline wplay/download_media.py:59]: Codacy Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163)

File: wplay/download_media.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    else:
+        await target_select.manual_select_target(page)
+
+    # Selectors
+    target_name_selector = "#main > header > div > div > div > span"
+    media_text = "#app > div > div > div > div > span > div > span > div > div > div > div > div > div > div > div > span"
+    media_images = "#app > div > div > div > div > span > div > span > div > div > span > div > div > div > div > div > div"
+    left_arrow_button = "#app > div > span > div > div > div > div > div > span"
+    media_url_img = "#app > div > span:nth-child(3) > div > div > div > div > div > div > div > div > img"
+    media_url_vid = "#app > div > span:nth-child(3) > div > div > div > div > div > div > div > div > video"
```
Comment (xandao6) [inline wplay/download_media.py:40]: we have a place to put the selectors, take a look in helpers.py

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- wplay/download_media.py  19
- wplay/__main__.py  1
         

Clones added
============
- wplay/download_media.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163&bid=17848965)

File: wplay/download_media.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    # Click on the `Media, Link and Docs` text
+    await page.waitForSelector(whatsapp_selectors_dict['media_text'])
+    await page.click(whatsapp_selectors_dict['media_text'])
+
+    # Click on the most recent media element
+    while True:
+        try:
+            await page.evaluate(f'''document.querySelector('{whatsapp_selectors_dict['media_images']}').click()''')
+            break
+        except:
```
Comment (rpotter12) [inline wplay/download_media.py:55]: Codacy Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163)

File: wplay/download_media.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                    if vid:
+                        vid = vid[-1]
+                        media_arr['vid'].append(vid)
+                        break
+            
+            # Go to next media element
+            await page.waitForSelector(whatsapp_selectors_dict['left_arrow_button'])
+            await page.evaluate(f'''document.querySelector('{whatsapp_selectors_dict['left_arrow_button']}').click()''')
+            time.sleep(0.5)
+
```
Comment (rpotter12) [inline wplay/download_media.py:94]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163)

File: wplay/download_media.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                    img = await page.evaluate(f'''() => [...document.querySelectorAll('{whatsapp_selectors_dict['media_url_img']}')]
+                                                        .map(element => element.src)''')
+                    if img and len(img) == 2:
+                        img = img[-1]
+                        if img not in media_arr:
+                            media_arr['img'].append(img)
+                        break
+                    countTry += 1
+                    time.sleep(0.3)
+            except:
```
Comment (rpotter12) [inline wplay/download_media.py:79]: Codacy Issue found: [No exception type(s) specified](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163)

File: wplay/download_media.py (hunk 6 above)
Comment (rpotter12) [reply wplay/download_media.py:55]: exception type is not written here

File: wplay/download_media.py (hunk 8 above)
Comment (rpotter12) [reply wplay/download_media.py:79]: No exception type is written here

File: wplay/download_media.py (hunk 7 above)
Comment (rpotter12) [reply wplay/download_media.py:94]: Remove white space from here

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- wplay/download_media.py  19
- wplay/__main__.py  1
         

Clones added
============
- wplay/download_media.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163&bid=17848965)

File: wplay/download_media.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                # If media is a video or gif
+                countTry = 0
+                while True:
+                    vid = await page.evaluate(f'''() => [...document.querySelectorAll('{whatsapp_selectors_dict['media_url_vid']}')]
+                                                        .map(element => element.src)''')
+                    if vid:
+                        vid = vid[-1]
+                        media_arr['vid'].append(vid)
+                        break
+
```
Comment (rpotter12) [inline wplay/download_media.py:89]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- wplay/download_media.py  19
- wplay/__main__.py  1
         

Clones added
============
- wplay/download_media.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5443163&bid=17848965)

**Pull request merged**